server/src/main/java/org/elasticsearch/indices/recovery/RecoveryState.java

```diff
@@ -147,7 +147,9 @@ public RecoveryState(StreamInput in) throws IOException {
     @Override
     public void writeTo(StreamOutput out) throws IOException {
         timer.writeTo(out);
-        out.writeByte(stage.id());
+        // We need to serialize the display stage so the correct
+        // stage is shown on the information APIs.
+        out.writeByte(getDisplayStage().id());
```

Comment:
This seems dangerous and I am not sure it is necessary. Any consumer providing the stage "to users" would have to use the `getDisplayStage` method instead and I think you already fixed those (did not check though)?

Reply:
I'm afraid that we need this change, `TransportRecoveryAction` extends `TransportBroadcastByNodeAction` that queries all the nodes that have the index shards allocated.

Reply:
I've added an extra field in `RecoveryState` so we serialize both values when possible.